core/module/digital_twin_graph/src/digital_twin_graph_impl.rs

```diff
@@ -0,0 +1,525 @@
+                    }
+                }
+                Ok(Err(error_message)) => {
+                    warn!("Failed to receive the answer request.  The error message is '{}'.  We may retry in a moment.", error_message);
+                    sleep(Duration::from_millis(Self::BACKOFF_BASE_DURATION_IN_MILLIS)).await;
+                    attempts_after_failure += 1;
+                    continue;
+                }
+                Err(error_message) => {
+                    warn!("Failed to receive the answer request.  The error message is '{}'.  We may retry in a moment.", error_message);
+                    sleep(Duration::from_millis(Self::BACKOFF_BASE_DURATION_IN_MILLIS)).await;
+                    attempts_after_failure += 1;
+                    continue;
+                }
+            }
+        }
+
+        debug!(
+            "Received an answer request.  The ask_id is '{}'. The payload is '{}",
+            answer_request.ask_id, answer_request.payload
+        );
+
+        Ok(tonic::Response::new(GetResponse { value: answer_request.payload.clone() }))
+    }
+
+    /// Set implementation.
+    ///
+    /// # Arguments
+    /// * `request` - Set request.
+    async fn set(
+        &self,
+        request: tonic::Request<SetRequest>,
+    ) -> Result<tonic::Response<SetResponse>, tonic::Status> {
+        warn!("Got a set request: {request:?}");
+
+        Err(tonic::Status::unimplemented("set has not been implemented"))
+    }
+
+    /// Invoke implementation.
+    ///
+    /// # Arguments
+    /// * `request` - Invoke request.
+    async fn invoke(
+        &self,
+        request: tonic::Request<InvokeRequest>,
+    ) -> Result<tonic::Response<InvokeResponse>, tonic::Status> {
+        let invoke_request = request.into_inner();
+        let instance_id = invoke_request.instance_id;
+        let member_path = invoke_request.member_path;
+        let request_payload = invoke_request.request_payload;
+
+        debug!("Received an invoke request for instance id {instance_id}");
+
+        // Retrieve the provider details.
+        let provider_endpoint_info_list = self
+            .find_digital_twin_providers_with_instance_id(
+                instance_id.as_str(),
+                digital_twin_protocol::GRPC,
+                &[digital_twin_operation::INVOKE.to_string()],
+            )
+            .await
+            .map_err(tonic::Status::internal)?;
+
+        if provider_endpoint_info_list.is_empty() {
+            return Err(tonic::Status::internal("No providers found"));
+        }
+
+        // We will only use the first provider.
+        let provider_endpoint_info = &provider_endpoint_info_list[0];
+
+        let provider_uri = provider_endpoint_info.uri.clone();
+        let instance_id = provider_endpoint_info.context.clone();
+
+        let tx = Arc::clone(&self.tx);
+        let mut rx = tx.subscribe();
+
+        let client_result = RequestClient::connect(provider_uri.clone()).await;
```

Comment:
there is quite a lot of duplicated code to send the ask request. we should probably abstract that out into a common method to avoid this duplication

Reply:
Fixed.